design/ceph-osd-removal.md

```diff
@@ -0,0 +1,113 @@
+# Removing OSDs from a Ceph cluster
+
+**Targeted for v1.X**
+
+## Background
+
+Removing OSDs is a common day-2 operation and users need to be able to perform
+this regularly.
+
+With the effort to add OSD replacement in the Ceph Dashboard, there is a demand
+for automating the deletion of individual OSDs.
+
+## Status
+
+As of Rook 1.1, there is some preliminary support for removing OSDs:
+
+### Removing Nodes
+
+Rook fully supports the removal of whole nodes from a Cluster, including all
+OSDs. This implementation does every call needed to completely remove all OSDs
+on that specific node from a Ceph cluster if that cluster no longer appears in
+the list of nodes in the CephCluster CR.
+
+There are concerns that this implementation might be dangerous, if a node is
+being removed inadvertently.
+
+### Removing individual OSDs
+
+Rook partially supports individual OSD-removal: The Rook operator monitors Ceph
+health and initiates the remove of OSDs that are, according to
+`ceph safe-to-destroy`, safe to be destroyed. Rook then removes the OSD
+Deployment automatically.
+
+Note that the device might stay referenced within the CephCluster CR.
+
+This feature is mainly aimed at automatically removing unresponsive OSDs.
+
+### Manual removal
+
+There is an existing Rook issue
+[#1827](https://github.com/rook/rook/issues/1827) containing a description of
+how to remove OSDs manually.
+
+## Design considerations
+
+Any future implementation within Rook needs to fulfill some requirements:
+
+1. To prevent inadvertent removal of OSDs, the act of removing OSDs needs to be
+   executed explicitly and manually.
+2. Omitting node or drive declarations within the CephCluster CR should not
+   physically remove OSDs.
+3. Except for physically removing data on disks, there should not be any manual
+   step required to allow a subsequent creation of a replacement OSD on the same
+   node.
+4. There should not be a perceived fundamental difference when removing whole
+   nodes versus removing single OSDs form a node.
+
+## Algorithm Outline
+
+Removing OSDs looks as follows when looking at the bare minimum of required
+steps (pseudocode):
+
+```python
+def ceph_crush_reweight(osd_id, weight): ...
+def ceph_pg_ls-by-osd(osd_id): ...
+def ceph_osd_out(osd_id): ...
+def ceph_osd_purge(osd_id): ...
+
+def remove_osd(osd_id, node, device_name):
+    ceph_crush_reweight(osd_id, 0.0)
+
+    while not len(ceph_pg-ls-by-osd(osd_id)):
+        time.sleep(10)
+
+    ceph_osd_out()
```

Comment:
For the implementation in the mgr module, could we get away with only these two steps?
- Set the CRUSH weight to 0.0
- Remove the OSD from the CephCluster CR

Could the operator take care of all the other steps after this? After the re-weighting is done this seems like it would work as long as the osd is marked as safe-to-destroy. This would mean we don't need any long-running operation in the mgr module. We would only need to know that the operator will delete the deployment (and remove it from the CRUSH map?) when it is safe-to-destroy.

Reply:
Looks like an OSD is `safe-to-destroy`, if it is either `out` and `down` or unused. so, I think this could work. And then, the operator would also call `ceph osd purge`?

Reply:
yes, i assume the operator could call purge when it's safe-to-destroy

Reply:
We might need to differentiate between remove and replace, but we can do that via the `destroyed` flag in the OSD

Reply:
Just pushed the change. looks good to me.